ouroboros-network/src/Ouroboros/Network/Server/ConnectionTable.hs

```diff
@@ -0,0 +1,212 @@
+
+instance Ord (ValencyCounter m) where
+    compare a b = compare (vcId a) (vcId b)
+
+instance Eq (ValencyCounter m) where
+    (==) a b = vcId a == vcId b
+
+-- | Returns current ValencyCounter value, represent the number of additional connections that
+-- can be created. May be negative.
+readValencyCounter :: MonadSTM m => ValencyCounter m -> STM m Int
+readValencyCounter vc = readTVar $ vcRef vc
+
+data ConnectionTableEntry m = ConnectionTableEntry {
+    -- | Set of ValencyCounter's for subscriptions interested in this peer.
+    cteRefs           :: !(Set (ValencyCounter m))
+    -- | Set of local SockAddr connected to this peer.
+    , cteLocalAddresses :: !(Set Socket.SockAddr)
+    }
+
+data ConnectionTableRef =
+    ConnectionTableCreate
+  -- ^ No connection to peer exists, attempt to create one.
+  | ConnectionTableExist
+  -- ^ A connection to the peer existed, either from another subscriber or the peer opened one
+  -- towards us.
+  | ConnectionTableDuplicate
+  -- ^ This subscriber already has counted a connection to this peer. It must try another target.
+  deriving Show
+
+-- | Add a connection.
+addValencyCounter :: MonadSTM m => ValencyCounter m -> STM m ()
+addValencyCounter vc = modifyTVar' (vcRef vc) (\r -> r - 1)
+
+-- | Remove a connection.
+remValencyCounter :: MonadSTM m => ValencyCounter m -> STM m ()
+remValencyCounter vc = modifyTVar' (vcRef vc) (+ 1)
+
+-- | Wait until ValencyCounter becomes positive, used for detecting when
+-- we can create new connections.
+waitValencyCounter :: MonadSTM m => ValencyCounter m -> STM m ()
+waitValencyCounter vc = do
+  v <- readTVar $ vcRef vc
+  when (v <= 0)
+    retry
+
+-- | Create a new ConnectionTable.
+newConnectionTable :: MonadSTM m => m (ConnectionTable m)
+newConnectionTable =  do
+    tbl <- newTVarM M.empty
+    li <- newTVarM 0
+    return $ ConnectionTable tbl li
+
+-- | Insert a new connection into the ConnectionTable.
+addConnection
+    :: forall m. (MonadSTM m)
+    => ConnectionTable m
+    -> Socket.SockAddr
+    -> Socket.SockAddr
+    -> Maybe (ValencyCounter m)
+    -- ^ Optional ValencyCounter, used by subscription worker and set to Nothing when
+    -- called by a local server.
+    -> STM m ()
+addConnection ConnectionTable{..} remoteAddr localAddr ref_m =
+    readTVar ctTable >>= M.alterF fn remoteAddr >>= writeTVar ctTable
+  where
+    fn :: Maybe (ConnectionTableEntry m) -> STM m (Maybe (ConnectionTableEntry m))
+    fn Nothing = do
+        refs <- case ref_m of
+                     Just ref -> do
+                         addValencyCounter ref
+                         return $ S.singleton ref
+                     Nothing -> return S.empty
+        return $ Just $ ConnectionTableEntry refs (S.singleton localAddr)
+    fn (Just cte) = do
+          let refs' = case ref_m of
+                           Just ref -> S.insert ref (cteRefs cte)
+                           Nothing  -> cteRefs cte
+          mapM_ addValencyCounter refs'
```

Comment:
I think it would be worth adding a comment line, that we update all valency counters for this particular peer.  This means that a single connection might be counted multiple times but for different `ValencyCounter`s, so they all must come from different `dns` names (or at least different calls to`dnsSubscriptionWorker`).

Reply:
I'll add a comment.